doc/api/diagnostics_channel.md

```diff
@@ -407,6 +455,507 @@ channel.subscribe(onMessage);
+
+The `start` event represents the synchronous start of a traceable function.
+At this point the event data may contain function arguments or anything else
+available at the very start of the execution of the function.
+
+#### `end(event)`
+
+* Name: `tracing:${name}:end`
+
+The `end` event represents the synchronous end of a traceable function. At this
+point, if the traced function was synchronous the `result` field will be set to
+the return value of the function. Alternatively, the `error` field may be
+present to represent any thrown errors.
+
+It is recommended to listen specifically to the `error` event to track errors
+as it may be possible for a traceable action to produce multiple errors. For
+example, an async task which fails may be started internally before the sync
+part of the task then throws an error.
+
+#### `asyncStart(event)`
+
+* Name: `tracing:${name}:asyncStart`
+
+The `asyncStart` event represents the callback or continuation of a traceable
+function being reached. At this point things like callback arguments may be
+available, or anything else expressing the "result" of the action.
+
+For callbacks-based functions, the first argument of the callback will be
+assigned to the `error` field, if not undefined, and the second argument will
+be assigned to the `result` field.
+
+For promises, the argument to the `resolve` path will be assigned to `result`
+or the argument to the `reject` path will be assign to `error`.
+
+It is recommended to listen specifically to the `error` event to track errors
+as it may be possible for a traceable action to produce multiple errors. For
+example, an async task which fails may be started internally before the sync
+part of the task then throws an error.
+
+#### `asyncEnd(event)`
+
+* Name: `tracing:${name}:asyncEnd`
+
+The `asyncEnd` event represents the callback of an async traceable function
+being completed. It's not likely event data will change after the `asyncStart`
+event, however it may be useful for some tracers to see the point where the
+callback completes.
+
+#### `error(event)`
+
+* Name: `tracing:${name}:error`
+
+The `error` event represents any error produced by the traceable function
+either synchronously or asynchronously. If an error is thrown in the
+synchronous portion of the traced function the error will be assigned to the
+`error` field of the event and the `error` event will be triggered. If an error
+is received asynchronously through a callback or promise rejection it will also
+be assigned to the `error` field of the event and trigger the `error` event.
+
+It is possible for a single traceable function call to produce errors multiple
+times so this should be considered when consuming this event. For example, if
+another async task is triggered internally which fails and then the sync part
+of the function then throws and error two `error` events will be emitted, one
+for the sync error and one for the async error.
+
 ### Built-in Channels
```

Comment:
The new channels should be added here

Reply:
Yep, will do when we finalize this. Need to wait for the ESM changes @GeoffreyBooth mentioned first. Thanks for the reminder though. :)

Reply:
Which changes in particular?

Reply:
The refactoring/merging stuff you mentioned. Not in a huge rush for this PR, it can wait for those changes.